rust/arrow/src/ffi.rs

```diff
@@ -0,0 +1,657 @@
+        array: *const FFI_ArrowArray,
+        schema: *const FFI_ArrowSchema,
+    ) -> Result<Self> {
+        if array.is_null() || schema.is_null() {
+            return Err(ArrowError::MemoryError(
+                "At least one of the pointers passed to `try_from_raw` is null"
+                    .to_string(),
+            ));
+        };
+        Ok(Self {
+            array: Arc::from_raw(array as *mut FFI_ArrowArray),
+            schema: Arc::from_raw(schema as *mut FFI_ArrowSchema),
+        })
+    }
+
+    /// creates a new empty [ArrowArray]. Used to import from the C Data Interface.
+    /// # Safety
+    /// See safety of [ArrowArray]
+    pub unsafe fn empty() -> Self {
+        let schema = Arc::new(FFI_ArrowSchema::empty());
+        let array = Arc::new(FFI_ArrowArray::empty());
+        ArrowArray { schema, array }
+    }
+
+    /// exports [ArrowArray] to the C Data Interface
+    pub fn into_raw(this: ArrowArray) -> (*const FFI_ArrowArray, *const FFI_ArrowSchema) {
+        (Arc::into_raw(this.array), Arc::into_raw(this.schema))
+    }
+
+    /// returns the null bit buffer.
+    /// Rust implementation uses a buffer that is not part of the array of buffers.
+    /// The C Data interface's null buffer is part of the array of buffers.
+    pub fn null_bit_buffer(&self) -> Option<Buffer> {
+        // similar to `self.buffer_len(0)`, but without `Result`.
+        let buffer_len = bit_util::ceil(self.array.length as usize, 8);
+
+        unsafe { create_buffer(self.array.clone(), 0, buffer_len) }
+    }
+
+    /// Returns the length, in bytes, of the buffer `i` (indexed according to the C data interface)
+    // Rust implementation uses fixed-sized buffers, which require knowledge of their `len`.
+    // for variable-sized buffers, such as the second buffer of a stringArray, we need
+    // to fetch offset buffer's len to build the second buffer.
+    fn buffer_len(&self, i: usize) -> Result<usize> {
+        let data_type = &self.data_type()?;
+
+        Ok(match (data_type, i) {
+            (DataType::Utf8, 1) => {
+                // the len of the offset buffer (buffer 1) equals length + 1
+                let bits = bit_width(data_type, i)?;
+                bit_util::ceil((self.array.length as usize + 1) * bits, 8)
+            }
+            (DataType::Utf8, 2) => {
+                // the len of the data buffer (buffer 2) equals the last value of the offset buffer (buffer 1)
+                let len = self.buffer_len(1)?;
+                // first buffer is the null buffer => add(1)
+                // we assume that pointer is aligned for `i32`, as Utf8 uses `i32` offsets.
+                #[allow(clippy::cast_ptr_alignment)]
+                let offset_buffer = unsafe {
+                    *(self.array.buffers as *mut *const u8).add(1) as *const i32
+                };
+                // get last offset
+                (unsafe { *offset_buffer.add(len / size_of::<i32>() - 1) }) as usize
```

Comment:
I spent several minutes thinking about whether the `- 1` is correct, seems it is, because `len` is the length of the offset buffer, which is one larger than the length of the corresponding array. Not sure how to phrase that in a small comment though.

Reply:
Thinking about this, an alternative is to safely cast all of this to i32 and perform the op on byte slice. IMO this is pretty unsafe if someone sends us data with wrong alignments. :/

Reply:
@jhorstmann , I agree with you. I am working on this on as a separate PR, but it is not forgotten.